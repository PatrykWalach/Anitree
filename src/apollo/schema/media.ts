import { ScoreFormat } from './viewer'
import { MediaList } from './mediaListCollection'

export interface Variables {
  isAdult?: boolean
  id: number
  format?: ScoreFormat
}

export interface Node {
  id: number
  type: MediaType
}

export type MediaType = 'MANGA' | 'ANIME'
export interface FuzzyDate {
  year: number | null
  month: number | null
  day: number | null
}

export interface Media extends Node {
  startDate: FuzzyDate
  endDate: FuzzyDate
  format: string | null
  description: string | null
  season: string | null
  seasonInt: number | null
  episodes: number | null
  chapters: number | null
  volumes: number | null
  isFavourite: boolean
  mediaListEntry: MediaList | null
  title: MediaTitle
  siteUrl: string
  relations: {
    edges: MediaEdge[]
  }
  tags: {
    rank: number
    name: string
  }[]
  studios: {
    nodes: Studio[]
  }
  status: MediaStatus | null
  bannerImage: string | null
  coverImage: MediaCoverImage
  stats: MediaStats
}
export interface MediaStats {
  scoreDistribution: ScoreDistribution[] | null
}
export interface ScoreDistribution {
  score: number
  amount: number
}
export interface MediaTitle {
  romaji: string
  english: string
  native: string
}

export type MediaRelation =
  | 'ADAPTATION'
  | 'PREQUEL'
  | 'SEQUEL'
  | 'PARENT'
  | 'SIDE_STORY'
  | 'CHARACTER'
  | 'SUMMARY'
  | 'ALTERNATIVE'
  | 'SPIN_OFF'
  | 'OTHER'
  | 'SOURCE'
  | 'COMPILATION'
  | 'CONTAINS'

export interface MediaEdge {
  node: Node
  relationType: MediaRelation
  id: number
}

export interface Studio {
  id: number
  name: string
}

export interface MediaCoverImage {
  extraLarge: string | null
  large: string | null
  medium: string | null
  color: string | null
}

export type MediaStatus =
  | 'CURRENT'
  | 'PLANNING'
  | 'COMPLETED'
  | 'DROPPED'
  | 'PAUSED'
  | 'REPEATING'
