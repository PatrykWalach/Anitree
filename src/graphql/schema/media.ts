import { ScoreFormat } from './viewer'
import { MediaList } from './mediaListCollection'

export interface Variables {
  isAdult?: boolean
  id: number
  format?: ScoreFormat
}

export interface Node {
  __typename: 'Media'
  id: number
  type: MediaType
}

export type MediaType = 'MANGA' | 'ANIME'
export interface FuzzyDate {
  __typename: 'FuzzyDate'
  year: number | null
  month: number | null
  day: number | null
}

export interface Media extends Node {
  startDate: FuzzyDate
  endDate: FuzzyDate
  format: string | null
  description: string | null
  season: string | null
  seasonInt: number | null
  episodes: number | null
  chapters: number | null
  volumes: number | null
  isFavourite: boolean
  ranknigs: MediaRank[]
  mediaListEntry: MediaList | null
  title: MediaTitle
  siteUrl: string
  relations: MediaConnection
  tags: MediaTag[]
  studios: StudioConnection
  status: MediaStatus | null
  bannerImage: string | null
  coverImage: MediaCoverImage
  stats: MediaStats
}
export interface MediaConnection {
  __typename: 'MediaConnection'
  edges: MediaEdge[]
}
export interface MediaTag {
  __typename: 'MediaTag'
  id: number
  rank: number
  name: string
}
export interface StudioConnection {
  __typename: 'StudioConnection'

  nodes: Studio[]
}
export interface MediaRank {
  __typename: 'MediaRank'
  allTime: boolean
  context: string
  rank: number
  type: 'RATED' | 'POPULAR'
  year: null | number
}
export interface MediaStats {
  __typename: 'MediaStats'
  scoreDistribution: ScoreDistribution[] | null
}
export interface ScoreDistribution {
  __typename: 'ScoreDistribution'
  score: number
  amount: number
}
export interface MediaTitle {
  __typename: 'MediaTitle'
  romaji: string
  english: string
  native: string
}

export type MediaRelation =
  | 'ADAPTATION'
  | 'PREQUEL'
  | 'SEQUEL'
  | 'PARENT'
  | 'SIDE_STORY'
  | 'CHARACTER'
  | 'SUMMARY'
  | 'ALTERNATIVE'
  | 'SPIN_OFF'
  | 'OTHER'
  | 'SOURCE'
  | 'COMPILATION'
  | 'CONTAINS'

export interface MediaEdge {
  __typename: 'MediaEdge'
  node: Node
  relationType: MediaRelation
  id: number
}

export interface Studio {
  __typename: 'Studio'
  id: number
  name: string
}

export interface MediaCoverImage {
  __typename: 'MediaCoverImage'
  extraLarge: string | null
  large: string | null
  medium: string | null
  color: string | null
}

export type MediaStatus =
  | 'CURRENT'
  | 'PLANNING'
  | 'COMPLETED'
  | 'DROPPED'
  | 'PAUSED'
  | 'REPEATING'
