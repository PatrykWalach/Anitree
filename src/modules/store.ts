import * as modules from '.'
import { Store, plugin } from 'vuex-composition-api'

export const store =
  process.env.NODE_ENV === 'production'
    ? undefined
    : new Store({
        plugins: [plugin(Object.values(modules))],
      })
