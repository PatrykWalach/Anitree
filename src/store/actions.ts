export { titleActions as title } from './modules/title'

export { editActions as edit } from './modules/edit'
export { filterActions as filter } from './modules/filter'
export { shareActions as share } from './modules/share'
export { settingsActions as settings } from './modules/settings'
export { timelineActions as timeline } from './modules/timeline'
export { changesActions as changes } from './modules/changes'
