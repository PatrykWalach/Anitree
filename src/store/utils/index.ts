export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
