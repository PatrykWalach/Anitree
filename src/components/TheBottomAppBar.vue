<template>
  <v-app-bar app bottom>
    <v-app-bar-nav-icon @click.stop="syncedDrawer = !syncedDrawer" />

    <v-tooltip
      top
      :key="title"
      v-for="{ icon, bind, title, on: actionOn } in mainActions"
    >
      <template v-slot:activator="{ attrs, on }">
        <v-btn
          v-bind="{ ...bind, ...attrs }"
          icon
          v-on="{ ...on, ...actionOn }"
        >
          <v-icon>{{ icon }}</v-icon>
        </v-btn>
      </template>
      <span>{{ title }}</span>
    </v-tooltip>

    <TheAppBarMenu :media="media" :actions="moreActions" />
    <TheFab top absolute />
  </v-app-bar>
</template>

<script lang="ts">
import { SetupContext, computed, createComponent } from '@vue/composition-api'
import TheFab, { useFab } from './TheFab.vue'
import { Media } from '../graphql/schema/media'
import { useMediaCardActions } from './MediaCardActions.vue'
import { useRoutes } from './TheAppBar.vue'
import { useSearch } from './TheFab.vue'

const TheAppBarMenu = () =>
  import(
    /* webpackChunkName: "TheAppBarMenu" */ /* webpackPrefetch: true */ './TheAppBarMenu.vue'
  )

export interface Props {
  media: Media | null
  drawer: boolean
}
export const useAppBarActions = (
  props: Pick<Props, 'media'>,
  root: SetupContext['root'],
) => {
  const { actions: _actions } = useMediaCardActions()

  const { active } = useFab(root)

  const { routeTitle } = useRoutes(root)

  const { searchBtn: _searchBtn } = useSearch()

  const searchBtn = computed(_searchBtn)
  const actions = computed(() => _actions(props.media))

  const allActions = computed(() =>
    (routeTitle.value ? actions.value : []).concat([searchBtn.value]),
  )

  const filteredActions = computed(() => {
    const { name } = root.$route
    const activeIcon = active.value.icon

    return allActions.value
      .filter(
        ({ icon, bind }) =>
          icon !== activeIcon && (!bind || !bind.to || bind.to.name !== name),
      )
      .reverse()
  })

  const mainActions = computed(() => filteredActions.value.slice().splice(0, 2))
  const moreActions = computed(() => filteredActions.value.slice().splice(2))

  return {
    allActions,
    filteredActions,
    mainActions,
    moreActions,
  }
}
export default createComponent<Readonly<Props>>({
  components: { TheAppBarMenu, TheFab },
  props: {
    drawer: { default: false, required: true, type: Boolean },
    media: { default: null, required: true, type: null },
  },
  setup(props, { root, emit }) {
    const syncedDrawer = computed({
      get: () => props.drawer,
      set: e => emit('update:drawer', e),
    })

    return {
      ...useAppBarActions(props, root),
      syncedDrawer,
    }
  },
})
</script>
