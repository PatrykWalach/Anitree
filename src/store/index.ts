import Vue from 'vue'
import Vuex from 'vuex'

// import { getModule } from 'vuex-module-decorators'
Vue.use(Vuex)

import { ModuleMedia } from './modules/media'
import { ModuleTheme } from './modules/theme'
import { ModuleTitle } from './modules/title'
import { ModuleAuth } from './modules/auth'

export interface RootState {
  media: ModuleMedia
  theme: ModuleTheme
  title: ModuleTitle
  auth: ModuleAuth
}

export default new Vuex.Store<RootState>({})
