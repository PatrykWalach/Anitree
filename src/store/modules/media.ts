import Vue from 'vue'
import { apolloQueryPage } from '../api'
import {
  Module,
  VuexModule,
  Mutation,
  Action,
  getModule,
  MutationAction
} from 'vuex-module-decorators'
import store from '@/store'
import {
  MediaNode,
  Media,
  MediaEdgeExtended,
  MediaEdge,
  Data,
  RawFilter,
  Filter,
  FetchVariables,
  Page
} from '@/types'

function createVariables(
  acc: {
    idIn: number[]
  }[],
  media: MediaNode
): {
  idIn: number[]
}[] {
  const index = acc.findIndex(arr => arr.idIn.length < 50)
  if (index === -1) {
    acc.push({ idIn: [media.id] })
  } else {
    acc[index].idIn.push(media.id)
  }
  return acc
}

function applyFilter(
  filteredEdges: MediaEdgeExtended[],
  filter: RawFilter
): MediaEdgeExtended[] {
  return filteredEdges.filter(filter)
}

function filterDuplicate<E = any>(
  test: (element: E, otherElement: E, i?: number, arr?: E[]) => boolean,
  element: E,
  i: number,
  arr: E[]
): boolean {
  return arr.findIndex(test.bind(null, element)) === i
}

function nodesAreEqual(element: MediaNode, otherElement: MediaNode) {
  return element.id === otherElement.id
}

function filterDuplicateNodes(
  node: MediaNode,
  i: number,
  allNodes: MediaNode[]
): boolean {
  return filterDuplicate(nodesAreEqual, node, i, allNodes)
}

function filterDuplicateEdges(
  edge: MediaEdge,
  i: number,
  allEdges: MediaEdge[]
): boolean {
  return filterDuplicateNodes(edge.node, i, allEdges.map(({ node }) => node))
}
@Module({
  namespaced: true,
  name: 'media',
  store,
  dynamic: true
})
export class ModuleMedia extends VuexModule {
  public currentId: number | null = null

  public media: Data<Media> = {}

  public filters: Data<Filter> = {}

  // public filteredMedia: MediaEdgeExtended[] = []

  public get activeFilters() {
    return Object.values(this.filters).filter(({ active }) => active)
  }

  public get exclusiveFilters(): RawFilter[] {
    return this.activeFilters
      .filter(({ exclusive }) => exclusive)
      .map(({ filter }) => filter)
  }

  public get inclusiveFilters(): RawFilter[] {
    return this.activeFilters
      .filter(({ exclusive }) => !exclusive)
      .map(({ filter }) => filter)
  }

  public get currentMedia(): Media | null {
    const { currentId, media } = this
    return (currentId && media[currentId]) || null
  }

  public get relationTypes(): string[] {
    const { inclusiveFilters, filteredMedia } = this
    return [
      ...new Set(
        inclusiveFilters
          .reduce(applyFilter, filteredMedia)
          .flatMap(media => media.node.relations.edges)
          .map(({ relationType }) => relationType)
      )
    ]
  }

  public get sortedMedia(): Media[] {
    const getYear = (seasonInt: number) => {
      const year = Math.floor(seasonInt / 10)
      return year > 50 ? 1900 + year : 2000 + year
    }

    return this.filteredMedia
      .map(({ node }) => node)
      .sort(
        (
          { startDate: dateA, seasonInt: seasonA },
          { startDate: dateB, seasonInt: seasonB }
        ) =>
          (seasonA &&
            seasonB &&
            (getYear(seasonA) - getYear(seasonB) ||
              (seasonA % 10) - (seasonB % 10))) ||
          (dateA.year && dateB.year && dateA.year - dateB.year) ||
          (dateA.month && dateB.month && dateA.month - dateB.month) ||
          (dateA.day && dateB.day && dateA.day - dateB.day) ||
          1
      )
  }

  public get filteredMedia(): MediaEdgeExtended[] {
    const getRelatedMedia = (
      edges: MediaEdge[],
      acc: MediaEdgeExtended[] = []
    ): MediaEdgeExtended[] => {
      const { exclusiveFilters, inclusiveFilters, media } = this

      const newMedia = exclusiveFilters.reduce(
        applyFilter,
        edges
          .map(({ relationType, node }) => {
            return {
              relationType,
              node: media[node.id]
            }
          })
          .filter(({ node }) => node)
      )

      acc.push(...newMedia)
      const accNodes = acc.map(({ node }) => node)
      const relatedNodes = inclusiveFilters
        .reduce(applyFilter, newMedia)
        .map(({ node }) => node)
        .flatMap(({ relations }) => relations.edges)
        .filter(filterDuplicateEdges)
        .filter(({ node }) => !accNodes.some(nodesAreEqual.bind(null, node)))

      if (relatedNodes.length) {
        return getRelatedMedia(relatedNodes, acc)
      }

      return acc
    }

    const { currentId } = this

    return currentId
      ? getRelatedMedia([
          {
            node: {
              id: currentId,
              type: ''
            },
            relationType: '',
            id: -1
          }
        ])
      : []
  }

  @Mutation
  public CHANGE_FILTER(filter: Filter): Filter {
    return Vue.set(this.filters, filter.id, filter)
  }

  @Mutation
  public ADD_MEDIA(media: Media): Media {
    return Vue.set(this.media, media.id, media)
  }

  @MutationAction
  public async CHANGE_CURRENT_ID({
    currentId
  }: {
    currentId: number | null
  }): Promise<{
    currentId: number | null
  }> {
    return { currentId }
  }

  @Action
  public async extendMediaEdge({
    relationType,
    node
  }: MediaEdge): Promise<MediaEdgeExtended> {
    return Promise.resolve({
      relationType,
      node: await this.getMedia(node.id)
    })
  }

  @Action
  public async fetchMedia(variables: FetchVariables): Promise<Media[]> {
    return this.fetchMediaPage(variables).then(page => page.media)
  }

  @Action
  public async fetchMediaPage(variables: FetchVariables): Promise<Page> {
    return new Promise(resolve =>
      apolloQueryPage(variables)
        .then(page => {
          const newMedia = page.media
          newMedia.forEach(this.ADD_MEDIA)
          return page
        })
        .then(resolve)
        .catch(() =>
          setTimeout(() => this.fetchMediaPage(variables).then(resolve), 10000)
        )
    )
  }

  @Action
  public async handleQueue(medias: Media[]): Promise<void> {
    return new Promise(resolve => {
      const { media } = this
      const newMedias: MediaNode[] = medias
        .flatMap(newMedia => newMedia.relations.edges)
        .map(({ node }) => node)
        .filter(filterDuplicateNodes)
        .filter(newMedia => !media[newMedia.id])

      if (newMedias.length) {
        const { fetchMedia, handleQueue } = this
        Promise.all(newMedias.reduce(createVariables, []).map(fetchMedia))
          .then(result => result.flat())
          .then(handleQueue)
          .then(resolve)
      } else {
        resolve()
      }
    })
  }

  // @Action
  // public async getFilteredMedia({
  //   edges = [
  //     {
  //       node: {
  //         id: this.currentId || -1,
  //         type: ''
  //       },
  //       relationType: '',
  //       id: -1
  //     }
  //   ],
  //   acc = []
  // }: {
  //   edges?: MediaEdge[]
  //   acc?: MediaEdgeExtended[]
  // }): Promise<MediaEdgeExtended[]> {
  //   return new Promise(async resolve => {
  //     const {
  //       exclusiveFilters,
  //       inclusiveFilters
  //       // ,media
  //     } = this

  //     const newMedia = exclusiveFilters.reduce(
  //       applyFilter,
  //       await Promise.all(edges.map(this.extendMediaEdge))
  //       // edges
  //       //   .map(({ relationType, node }) => {
  //       //     return {
  //       //       relationType,
  //       //       node: media[node.id]
  //       //     }
  //       //   })
  //       //   .filter(({ node }) => node)
  //     )

  //     acc.push(...newMedia)
  //     const accNodes = acc.map(({ node }) => node)
  //     const relatedNodes = inclusiveFilters
  //       .reduce(applyFilter, newMedia)
  //       .map(({ node }) => node)
  //       .flatMap(({ relations }) => relations.edges)
  //       .filter(filterDuplicateEdges)
  //       .filter(({ node }) => !accNodes.some(nodesAreEqual.bind(null, node)))

  //     if (relatedNodes.length) {
  //       this.getFilteredMedia({ edges: relatedNodes, acc }).then(resolve)
  //     } else {
  //       resolve(acc)
  //     }
  //   })
  // }

  // @Mutation
  // public CHANGE_FILTERED_MEDIA(filteredMedia: MediaEdgeExtended[]) {
  //   this.filteredMedia = filteredMedia
  // }

  // @Action({ commit: 'CHANGE_FILTERED_MEDIA' })
  // public async changeFilteredMedia() {
  //   return await this.getFilteredMedia({})
  // }

  @Action
  public async getMedia(id: number) {
    return new Promise<Media>(async (resolve, reject) => {
      const storedMedia = this.media[id]
      if (storedMedia) {
        resolve(storedMedia)
      } else {
        const media = await this.fetchMedia({
          idIn: [id]
        })

        if (media.length) {
          resolve(media.shift())
        } else {
          reject()
        }
      }
    })
  }
}
export const media =getModule(ModuleMedia)
export default media
