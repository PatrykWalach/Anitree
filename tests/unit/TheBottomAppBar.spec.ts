import { createVariations, matchSnapshot } from './utils'
import TheBottomAppBar from '@/components/TheBottomAppBar.vue'
import { mockedMedia } from './mocks/media'
import { useMockedStore } from './mocks/store'

describe('TheBottomAppBar', () => {
  createVariations({
    mocks: { $route: { name: () => ['home', 'title-timeline'] } },
    propsData: {
      drawer: () => [true, false],
      media: () => [mockedMedia, null],
    },
  }).forEach(props =>
    matchSnapshot(TheBottomAppBar, {
      ...props,
      ...useMockedStore(),
    }),
  )
})
