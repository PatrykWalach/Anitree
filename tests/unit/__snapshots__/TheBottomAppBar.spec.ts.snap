// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheBottomAppBar matches snapshot 1`] = `
<v-app-bar-stub
  app="true"
  bottom="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-app-bar-nav-icon-stub />
    
  <!---->
   
  <thefab-stub
    absolute=""
    top=""
  />
</v-app-bar-stub>
`;

exports[`TheBottomAppBar matches snapshot 2`] = `
<v-app-bar-stub
  app="true"
  bottom="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-app-bar-nav-icon-stub />
    
  <!---->
   
  <thefab-stub
    absolute=""
    top=""
  />
</v-app-bar-stub>
`;

exports[`TheBottomAppBar matches snapshot 3`] = `
<v-app-bar-stub
  app="true"
  bottom="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-app-bar-nav-icon-stub />
    
  <!---->
   
  <thefab-stub
    absolute=""
    top=""
  />
</v-app-bar-stub>
`;

exports[`TheBottomAppBar matches snapshot 4`] = `
<v-app-bar-stub
  app="true"
  bottom="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-app-bar-nav-icon-stub />
    
  <!---->
   
  <thefab-stub
    absolute=""
    top=""
  />
</v-app-bar-stub>
`;

exports[`TheBottomAppBar matches snapshot 5`] = `
<v-app-bar-stub
  app="true"
  bottom="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-app-bar-nav-icon-stub />
    
  <!---->
   
  <thefab-stub
    absolute=""
    top=""
  />
</v-app-bar-stub>
`;

exports[`TheBottomAppBar matches snapshot 6`] = `
<v-app-bar-stub
  app="true"
  bottom="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-app-bar-nav-icon-stub />
    
  <!---->
   
  <thefab-stub
    absolute=""
    top=""
  />
</v-app-bar-stub>
`;

exports[`TheBottomAppBar matches snapshot 7`] = `
<v-app-bar-stub
  app="true"
  bottom="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-app-bar-nav-icon-stub />
    
  <!---->
   
  <thefab-stub
    absolute=""
    top=""
  />
</v-app-bar-stub>
`;

exports[`TheBottomAppBar matches snapshot 8`] = `
<v-app-bar-stub
  app="true"
  bottom="true"
  extensionheight="48"
  src=""
  tag="header"
  tile="true"
  value="true"
>
  <v-app-bar-nav-icon-stub />
    
  <!---->
   
  <thefab-stub
    absolute=""
    top=""
  />
</v-app-bar-stub>
`;
